openlcs/libs/corgi.py

```diff
@@ -238,3 +276,94 @@ def get_paginated_data(self, query_params=None, api_path="components"):
+            return (False, component.get("purl"))
+        link = find_srpm_source(sources)
+        if link:
+            logger.info("Querying component: %s", link)
+            component = self.get(link)
+            return (True, component) if component else (False, link)
+        else:
+            return (False, component.get("purl"))
+
+    def fetch_component(self, link):
+        component = self.get(link)
+        if component:
+            if component.get("type") == "RPM":
+                return self.get_srpm_component(component)
+            # FIXME: examine what to return for non-rpm components.
+            return (True, component)
+        else:
+            return (False, link)
+
+    def get_container_source_components(self, component):
+        """
+        Extract source components from a corgi container component
+
+        If the name of the component ends with "-source", it will be handled
+        separately. Otherwise the function looks for binary components
+        provided by the component and retrieves their corresponding
+        source components. The source components are returned as a list.
+        There are chances that after several retries the component retrieval
+        still fails, failed ones are returned as a list of strings.
+
+        Args:
+        component (dict): The OCI component info
+
+        Returns:
+        Tuple: (components found / missings)
+        """
+        name = component.get("name")
+        if name.endswith("-source"):
+            logger.debug("This is a source container build")
+            sources = component.get("sources")
+            if not sources:
+                return None
+            link = sources[0].get("link")
+            component = self.get(link)
+            logger.debug("Binary build %s retrieved.", component['nevra'])
+
+        oci_provides = component.get("provides", [])
+        links = [provide.get("link") for provide in oci_provides]
+        logger.debug("List of provides(%d) collected", len(links))
+
+        components_extracted = list()
+        components_missing = set()
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=3
+        ) as executor:
+            tasks = {
+                executor.submit(
+                    self.fetch_component, link): link for link in links}
+            for task in concurrent.futures.as_completed(tasks):
+                success, result = task.result()
+                if success:
+                    components_extracted.append(result)
+                    logger.debug("Provides %s retrieved.", result['nevra'])
+                else:
+                    components_missing.add(result)
+                    logger.error("Provides %s missing.", result)
+        # Remove duplicates
+        component_set = set(tuple(c.items()) for c in components_extracted)
+        components = [dict(c) for c in component_set]
+        return (components, list(components_missing))
```

Comment:
`return components, list(components_missing)`

Reply:
These are equivalent. I prefer using `()` so that it's immutable.

Reply:
a comma makes a tuple. thus 2 terms are equivalent. see also https://docs.python.org/3/library/stdtypes.html#tuple

When unpacking, there's nothing different.

Reply:
ok, I don't mean incorrect, but it's unnecessary to add